EntityMask should be an enum class

rniwa opened the pull request:
#### d827efd1ab4ac43758f5decd5abcace52d2743b9
<pre>
EntityMask should be an enum class
<a href="https://bugs.webkit.org/show_bug.cgi?id=257305">https://bugs.webkit.org/show_bug.cgi?id=257305</a>

Reviewed by Wenson Hsieh.

* Source/WebCore/editing/MarkupAccumulator.cpp:
(WebCore::appendCharactersReplacingEntitiesInternal):
(WebCore::MarkupAccumulator::appendCharactersReplacingEntities):
(WebCore::MarkupAccumulator::entityMaskForText const):
(): Deleted.
* Source/WebCore/editing/MarkupAccumulator.h:

Canonical link: <a href="https://commits.webkit.org/264769@main">https://commits.webkit.org/264769@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/72fbb5a99913f99b8470691df49778b4514b3f45

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/8539 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/8829 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/9043 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/10194 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/8567 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/8548 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/10818 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/8793 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/11435 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/8685 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/9714 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/7680 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/10350 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/7014 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/7809 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/15352 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/8133 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/7958 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/11305 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/8434 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/6897 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/7711 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/11920 "Built successfully") | | | 
| [❌ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/1011 "Failed to push commit to Webkit repository") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/8176 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebCore/editing/MarkupAccumulator.cpp:
```diff
@@ -56,51 +56,51 @@ using namespace HTMLNames;
-    { "&#9;", 4, EntityTab },
-    { "&#10;", 5, EntityLineFeed },
-    { "&#13;", 5, EntityCarriageReturn },
+    { "", 0 , static_cast<EntityMask>(0) },
```
Would it make sense to have a sentinel constant like so?

```
constexpr EntityMask EntityMaskNull = static_cast<EntityMask>(0);
```

whsieh replied:
I think that's an interesting point — is it technically undefined behavior to static cast a raw int to an invalid `enum class`?

(Perhaps we could avoid this altogether by making `EntityDescription`'s `mask` an `OptionSet<EntityMask>` and then just pass `{ }` here).

rniwa replied:
Yeah, I've considered making it OptionSet as well. Maybe that's cleaner?

rniwa replied:
Actually, using std::optional<> makes more sense because these mask values refer to a specific entity type, not a mask of entities.

rniwa committed:
EntityMask should be an enum class
https://bugs.webkit.org/show_bug.cgi?id=257305

Reviewed by Wenson Hsieh.

* Source/WebCore/editing/MarkupAccumulator.cpp:
(WebCore::appendCharactersReplacingEntitiesInternal):
(WebCore::MarkupAccumulator::appendCharactersReplacingEntities):
(WebCore::MarkupAccumulator::entityMaskForText const):
(): Deleted.
* Source/WebCore/editing/MarkupAccumulator.h:

Canonical link: https://commits.webkit.org/264769@main
